File: sort/sort-integers/python/IntArray.py

import sys

class IntArray:
    """Represents an array of integers, which can be sorted, etc."""

    def __int__(self):
        """An array of integers, which can be sorted, etc."""
        None

    @staticmethod
    def sort(input: [int]) -> [int]:
        """Sorts an input array of integers, in O(n) time, where n is the length of the array."""

        # Define the range of values in the array.
        low = sys.maxsize
        high = -1 * (sys.maxsize - 1)
        for value in input:
            if value < low:
                low = value
            if value > high:
                high = value
        #print(f"low: {str(low)}; high: {str(high)}." )

        # Collect the count of each value in an array indexed by the value.
        buckets = (high-low+1) * [0]
        for value in input:
            buckets[value - low] += 1
        #print(buckets)

        result = len(array) * [0] # Allocate result list to match size of input.
        index = 0
        for value_shifted in range(0, len(buckets)):
            for k in range(0, buckets[value_shifted]):
                result[index] = value_shifted + low
                index += 1

        return result


if __name__ == "__main__":
    array = [1, 499, 101, 52, 81, 237, 202, 311, 195, 16]

    result = IntArray.sort(array)

    print ("Input: " + " ".join(str(n) for n in array).rstrip())
    print ("Sorted: " + " ".join(str(n) for n in result).rstrip())

File: sort/sort-integers/csharp/IntArray.cs

// Given an array of integers with values of the elements in the range [1:500], sort the array.
// You can test also at: https://dotnetfiddle.net/

using System;

namespace IntArray {

    public class IntArray {

        public static void Main(string[] args) {
            int[] input = new int[10]
            {
                1, 499, 101, 52, 81, 237, 202, 311, 195, 16
            };
            
            int[] sorted = IntArray.Sort(input);

            string s = "";
            foreach (int x in sorted)
            {
                s += x.ToString() + " ";
            }

            System.Console.WriteLine("Sorted: " + s);
        }
        
        // Sorts an array of integers falling within a range (low to high).
        // Returns a new array with the values sorted.
        // Runs in O(n), where n is the size of the input array.
        public static int[] Sort(int[] input)
        {
            // Find the range of values.
            int low = 1;
            int high = 500;

            int[] buckets = new int[high - low + 1];

            // Collect the count of each value in an array indexed by the value.
            foreach (int i in input)
            {
                // Value i is shifted by an offset of low, to fit in the array.
 				//Console.WriteLine("{0}, {1}, {2}", i, i-low, buckets.Length);
                buckets[i - low]++;
            }
            
            int[] result = new int[input.Length];
            int index = 0;
            for (int n=0; n < buckets.Length; n++)
            {
                // The inner loop includes values only if their count > 0.
                for (int j=0; j < buckets[n]; j++)
                {
                    // The value is the offset index k plus the shift amount, which is low.
					//Console.WriteLine("{0}, {1}, {2}", n + low, index, result.Length);                    
                    result[index] = n + low;
                    index++;
                }
            }
            
            return result;
        }
    }
}

